TChannel + Hyperbahn + Ringpop + Thrift scaffold

Raynos opened the pull request:
This is a first pass at a scaffold for tchannel and pals.

Features:
- ringpop out of the box
- hyperbahn out of the box
- thrift out of the box; including examples of thrift exceptions
- 100% code coverage
- 100% integration tests
- All tests go through hyperbahn client by importing `TestHyperbahnCluster` from tchannel
- Zero folders 

It may help to read:
- [tchannel guide](https://github.com/uber/tchannel/blob/master/node/GUIDE.md)
- [tchannel docs](https://github.com/uber/tchannel/tree/master/node/docs)

r: @Matt-Esch @rf @jcorbin @jwolski

cc @willyham @shannili

Raynos committed:
general cleanup

Raynos committed:
an idea

Raynos committed:
get template ready for local registration

Raynos committed:
get rid of some of the OSS concerns

Raynos committed:
get health test passing

Raynos committed:
fix some typoes and stuff

Raynos committed:
suppport the child-process test

Raynos committed:
get that 100% coverage

Raynos committed:
leave some TODOs

Willyham commented on template/content/endpoints.js:
```diff
@@ -0,0 +1,62 @@
+    );
+}
+
+Endpoints.prototype.health =
```
I'd suggest adding an `explode` example because throwing exceptions correctly is nuanced and everyone should be doing it right.

Willyham replied:
Ignore, just seen below.

Willyham commented on template/content/test/child-process.js:
```diff
@@ -0,0 +1,79 @@
+        '--port', '0'
+    ];
+
+    var hostPortList = cluster.hyperbahnCluster.hostPortList;
```
This seems like it would be easier to import your server as a module and call it rather than spawning a new process. Wouldn't you want a fatal exception when running the tests to fail fast and exit?

Willyham replied:
Just realised that this is what you're testing. Why use TestCluster at all for this? Maybe I'm missing the point of test cluster

Raynos replied:
It uses TestCluster with appCount: 0 to get `cluster.hyperbahnCluster` and `cluster.client`

This means it tells the child process to advertise with hyperbahn and then uses `cluster.client` to talk to the child process through hyperbahn.

The `TestCluster` does three things
- spins up a hyperbahn ring of size 2
- creates your `test-client.js` which talks to hyperbahn
- spins up an app ring of size N.

Willyham commented on template/content/test/child-process.js:
```diff
@@ -0,0 +1,79 @@
+        assert.ok(lines.length > 0);
+
+        var startedLine = lines.filter(function find(x) {
+            return x.indexOf('server started') >= 0;
```
Seems a little brittle, at the very least a comment where this is logged to denote that this is not an arbitrary string. Is the health check below not enough?

Raynos replied:
will leave comment.

Willyham commented on template/content/test/lib/test-cluster.js:
```diff
@@ -0,0 +1,105 @@
+    self.client = null;
+}
+
+TestCluster.prototype.bootstrap = function bootstrap(cb) {
```
Where is this called? Can't find it anywhere

rf replied:
its called by tape-cluster

rf commented on template/content/app.js:
```diff
@@ -0,0 +1,44 @@
+    options = options || {};
+
+    self.config = fetchConfig(__dirname, {
+        dcValue: 'todo',
```
isnt this just gonna be a quick readfilesync or whatever? just do it

Raynos replied:
The filepath is company specific.

rf commented on template/content/app.js:
```diff
@@ -0,0 +1,44 @@
+        logger: options.logger,
+        statsd: options.statsd
+    });
+    self.endpoints = Endpoints(self);
```
honestly don't like this pattern, endpoints shouldn't need the whole app object, just pass it what it needs, preferably the channel and the thrift 'codec'

jwolski replied:
anything that has a reference to `app` will have direct access to the endpoint handlers. Attaching the `endpoints` obj to the app instance is a bit weird and I think it'll lead to confusion for developers.

Raynos replied:
ill collapse endpoints and app into one thing

rf commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+var SERVICE_NAME = 'my-service';
+
+/* TODO:
+    - process reporter
```
https://github.com/rf/process-reporter

Raynos replied:
:+1:

rf commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+    path.join(__dirname, 'thrift', 'service.thrift'), 'utf8'
+);
+
+var SERVICE_NAME = 'my-service';
```
why not loaded out of config

jwolski replied:
I agree. I think every value that _needs_ to be replaced by the developer should be called out explicitly. Centralize all of those in configs or documentation, but not dispersed as both hard-coded values AND config.

Raynos replied:
will move more stuff into config

rf commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+        backupFile: config.get('clients.uncaught-exception.backupFile')
+    });
+    self.bootFile = config.get('bootFile');
+    self.port = config.get('port');
```
how are you going to have a different port per ringpop node

Raynos replied:
`--port` in supervisor.ini

rf commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+    self.hyperbahnClient = HyperbahnClient({
+        tchannel: self.rootChannel,
+        serviceName: SERVICE_NAME,
+        hostPortList: config.get('clients.hyperbahn.seedList'),
```
do it right and read from /etc/ for prod

Raynos replied:
Yes I need to add file reading support; will do.

I do want the hard coded path to be passed in as cli args though; cant hard code `/etc/autobahn` here.

rf commented on template/content/endpoints.js:
```diff
@@ -0,0 +1,62 @@
+    var thrift = app.clients.tchannelThrift;
+
+    thrift.register(
+        channel, 'MyService::health_v1', app, self.health
```
I'd rather it be anonymous functions that 'bind' the methods on the prototype

Raynos replied:
Fixed; by moving into app.js as static methods

rf commented on template/content/endpoints.js:
```diff
@@ -0,0 +1,62 @@
+}
+
+Endpoints.prototype.health =
+function health(app, req, head, body, cb) {
```
If a method is being called with a `this` that isn't bound to the `Endpoints` instance it shouldn't be on the prototype. It doesn't make sense for it to be on the prototype. Obviously it works but it's confusing and unnecessary. Just bind `this` correctly like everyone else

Willyham replied:
+1 in a real app these are likely going to be too complex to have in one file as well which would make this pattern quite strange

jwolski replied:
Agreed that it's unlikely that devs will put all handler implementations in a single file. @Willyham are you thinking about one file per endpoint handler?

Raynos replied:
yes this pattern of methods on the prototype without `this` is silly; will fix.

I'm going to move all endpoints into app and collapse the two.

Willyham commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+            return cb(err);
+        }
+
+        self.hyperbahnClient.once('advertised', cb);
```
You're going to get shit for spelling this correctly :gb:

Willyham commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+        }
+
+        self.hyperbahnClient.once('advertised', cb);
+        self.hyperbahnClient.advertise();
```
Do you catch the error from this anywhere? I think the process is not going to quit if there's an error advertising

Raynos replied:
I'll add an error handler

jwolski commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+        statsd: self.statsd
+    });
+
+    self.ringpop = Ringpop({
```
As you may know, Ringpop has many parameters (like timeouts, protocol period durations, etc) that can be specified by the application. I think that it'd be nice to put those provisions here so that these defaults can be extended with user-specified config, for example:

```
Ringpop(_.defaults({
  app: SERVICE_NAME,
  // etc
}), config.get('clients.ringpop.ctorOpts'));
```

Raynos replied:
people can just add more paramaters; this file is your application; if you want to add an option just edit this file.

jwolski replied:
Yep. That's fine with me.

jwolski commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+
+    self.logger = options.logger || Logtron({
+        meta: {
+            team: 'my-team',
```
`my-team`: move to config?

Raynos replied:
+1

jwolski commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+*/
+module.exports = ApplicationClients;
+
+function ApplicationClients(config, options) {
```
This ctor is already busy. It'll only get busier once they add their client dependencies. Have you thought about ways to decompose this file into smaller bits? One can go too far with this idea too, so I understand if you want to keep it monolithic and simple.

Raynos replied:
This is a pretty busy thing; totally agree;

However I tried separating it out and its just spreads the mess around. I think keeping it all in one monolithic file is a good idea as this is really the entire glue layer.

Spreading the complexity out just hides it.

Matt-Esch replied:
I am working on a project with a monster number of clients. I will be decomposing it into a dependency graph by  exporting each client with a list of dependencies and then initializing everything with run-auto. This way I can keep client initialization code factored out into simple functions without needing to maintain a huge order-dependent initialization script.

jwolski commented on template/content/server.js:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env node
+'use strict';
```
This module doesn't do much and in fact, I think it leads to some initial confusion about what the distinction between app.js and server.js are.

Raynos replied:
this module is the entry point; 

`node server.js` starts the application `app.js` is something you can import in your tests.

I can add more "features" to this file.

jwolski commented on template/content/test/lib/test-cluster.js:
```diff
@@ -0,0 +1,105 @@
+    self.bootFile = null;
+
+    self.hyperbahnCluster = HyperbahnCluster({
+        size: 2
```
make size specifiable/overridable

EDIT: nevermind. i didnt at first realize it was the _hyperbahn_ cluster. Specifying this size is less important now.

jwolski commented on template/content/app.js:
```diff
@@ -0,0 +1,44 @@
+    self.endpoints = Endpoints(self);
+
+    // Example data structure on application
+    self.exampleDb = {};
```
Im in favor of getting rid of `exampleDb`. If this generator is the scaffolding meant to be built on top of, every developer is going to have to remove this line of code from their app. I think that's wasted time. And this `exampleDb` is best left for an "example app".

Raynos replied:
I dont know whether the scaffold and the example app should be seperate.

I basically scaffolded out some simple endpoints to demonstrate how to do testing. Without those endpoints it's hard to demonstrate testing with just the health endpoint.

Previous scaffolds didn't have enough tests and left it open for developers to write their own tests. I want this scaffold to have some solid examples for how to write integration tests.

jwolski commented:
I think this looks good. Some high-level comments/remarks:
- I think we have to consider how much "sample" code we want in this template. If this template will serve as the scaffolding for all future hyperbahn/tchannel-based projects then there's a little too much sample code in this template (aka the endpoints handlers get + put) and will have to be deleted every time. You should be careful to reduce that as much as possible.
- Do we want to make this more modular or not? e.g. a endpoint handler folder instead of a single file. I know, too much modularity sucks.
- It's not immediately obvious to me the intended purpose and differences between server.js and app.js.

jwolski commented on template/content/clients.js:
```diff
@@ -0,0 +1,141 @@
+'use strict';
```
Might be worth considering renaming this to `deps.js`. "Clients" gets a bit ambiguous especially with things like loggers, etc. You typically dont think about your logger as a "client" to the logging subsystem, but you do think of a logger as a dependency of your service. Food for thought...

Raynos replied:
Actually; I want the logger to be a service over hyperbahn soon (tm).

I'll stick to `clients` for now in the term of "clients to other services and pieces of infrastructure"

Raynos committed:
leave a comment in child-process.js

Raynos committed:
improve README & package

Raynos committed:
fix linting

Raynos committed:
configure linting and such

Raynos committed:
configure linting for template

Raynos committed:
collapse endpoints into app

Raynos committed:
clean up linting

Raynos committed:
add process-reporter

Raynos committed:
upgrade latest version of nyc

Raynos committed:
server exit 1 if we cannot start

Raynos committed:
move serviceName into config

Raynos committed:
get team from config

Raynos committed:
handle hyperbahnClient better

Raynos committed:
store team in config

Raynos committed:
add ignore comments + deps

Raynos committed:
server: print extra details in startup

Raynos committed:
server: add startup time stat

Raynos committed:
refactor child-process code into helper funciton

Raynos committed:
child-process: test that processTitle feature

Raynos committed:
set correct flag

Raynos committed:
leave a bunch of TODO delete me comments

Raynos commented:
@jwolski @Willyham @rf addressed all feedback

Raynos commented:
@jwolski 

> I think we have to consider how much "sample" code we want in this template. If this template will serve as the scaffolding for all future hyperbahn/tchannel-based projects then there's a little too much sample code in this template (aka the endpoints handlers get + put) and will have to be deleted every time. You should be careful to reduce that as much as possible.

I added a bunch of "TODO remove me" comments to the example code and grouped it so that it's as easy to delete as possible.

I'm leaving it there to demonstrate how to write tests using `test-client.js` as that is not something documented anywhere else (how to implement endpoints and write thrift files is documented in the tchannel GUIDE.md).

> Do we want to make this more modular or not? e.g. a endpoint handler folder instead of a single file. I know, too much modularity sucks.

What we found with the autobahn service is that breaking things into more folders & files creates a structure around the "framework" rather then a structure around your application. This time round I want to encourage less structure and files and have the application structure occur organically from the actual application code you write.

As people implement their service they should break it out into classes that logically belong to the service. As an example see tcollector which has an Aggregator and Span class.

> It's not immediately obvious to me the intended purpose and differences between server.js and app.js.

I added more features to `server.js` so it should be more obvious now.

jwolski commented:
lgtm

Willyham commented on template/content/.eslintrc:
```diff
@@ -0,0 +1,6 @@
+{
```
Bit of an aside, but if we're changing it in a scaffold shouldn't we consider changing it in the standard?

Raynos replied:
We should. I'll open a pr

Matt-Esch replied:
:+1:

Willyham commented:
:+1: Looks good to me. I still think there's too much code which could go into a library though :trollface:

rf commented:
:+1:

Matt-Esch commented on .eslintrc:
```diff
@@ -0,0 +1,6 @@
+{
```
We should really just make this the standard, I have to put this in every project I touch and it's a necessity to write efficient code.

Raynos replied:
These two rules have been upstreamed into uber-standard. I'll remove this file.

Matt-Esch commented on template/content/package.json:
```diff
@@ -17,55 +17,35 @@
   ],
   "dependencies": {
+    "git-repo-info": "1.1.2",
+    "logtron": "8.3.4",
```
You want to update this to prune the deps I removed from logtron

Matt-Esch commented on template/content/app.js:
```diff
@@ -0,0 +1,94 @@
+    self.clients.destroy();
+};
+
+Application.health = function health(app, req, head, body, cb) {
```
Do you really suggest putting methods on Application in a production service? I think it's bad to encourage people to use functions like objects for the sake of squeezing all of this in here.

Raynos replied:
I previously had an `endpoints.js` which didn't feel quite right. In autobahn i have an endpoints folder but that feels quite heavy for simple services.

I think this might actually be the right balance; your endpoints should be small and you will only have a couple of them; put them in here and if they get too big move the code out of the endpoints function.

Matt-Esch commented on template/content/package.json:
```diff
@@ -17,55 +17,35 @@
-    }
-  },
-  "private": true
+  }
```
seems to be missing an opinion about npm/shrinkwrap

Raynos replied:
shrinkwrap is out of scope for this scaffold.

Matt-Esch commented on template/content/server.js:
```diff
@@ -0,0 +1,69 @@
+
+    var bootstrapEnd = Date.now();
+
+    function onAppReady(err) {
```
Poorly named. This is onClientsReady or onDepsReady. Consider moving the closure outside, and passing `(err, opts)` to it with a smaller closure.

Matt-Esch commented on template/content/server.js:
```diff
@@ -0,0 +1,69 @@
+            return process.exit(1);
+        }
+
+        var processTitle = opts.argv && opts.argv.processTitle;
```
I don't think you should wait until this point to set the process title. You want to do it as one of the first things as process monitoring often depends on process title.

Matt-Esch commented on template/content/clients.js:
```diff
@@ -0,0 +1,170 @@
+module.exports = ApplicationClients;
+
+function ApplicationClients(config, options) {
+    /*eslint no-process-env: 0*/
```
I'd rather you deleted this and put `var NODE_ENV =  require('process').env.NODE_ENV;` at the top.

Raynos committed:
update dev dependencies

Raynos committed:
update prod dependencies

Raynos committed:
server: more process.title and better failure mode

Raynos committed:
add missing istanbul ignore

Raynos commented:
Addressed all feedback.
